lib/src/main/java/io/ably/lib/transport/ConnectionManager.java

```diff
@@ -1232,12 +1230,43 @@ private synchronized void onConnected(ProtocolMessage message) {
             requestState(transport, new StateIndication(ConnectionState.failed, e.errorInfo));
             return;
         }
-
         /* indicated connected currentState */
         setSuspendTime();
         requestState(new StateIndication(ConnectionState.connected, error));
     }
 
+    /**
+     * Add all pending queued messages to the front of QueuedMessages for them to be sent later
+     * Spec: RTN19a
+     * @param resetMessageSerial whether to reset message serial, this will determine whether to reset message serials
+     * on pending queue, for example when a connection resume failed
+     */
+    private void addPendingMessagesToQueuedMessages(boolean resetMessageSerial) {
+        //RTN19a
+        if (resetMessageSerial){
+            pendingMessages.resetStartSerial(0);
+            msgSerial = 0; //msgSerial will increase in sendImpl when messages are sent
+        }
+        final List<QueuedMessage> messagesFromPendingQueue = CollectionUtils.deepCopyOfQueuedMessages(pendingMessages.queue);
+        //Clear the pendingQueue now, because we do not want the retried messages to accumulate on it.
+        pendingMessages.clearQueue();
+
+        //put messages from pending messages to front of queuedMessages, the ones with the message serials will already
+        //have been assigned new message serial to them at this point
+        queuedMessages.addAll(0,messagesFromPendingQueue);
+        // reassign new serials for remaining queued messages if reset was required
+        if (resetMessageSerial) {
```

Comment:
Why reset the `msgSerial` as two separate operations (ie once for the contents of `pendingMessages`, and once for the original contents of `queuedMessages`. Is it simpler just to add the messages from `pendingMessages`, and then set the new `msgSerial` for all of them as a single operation?